rmw_fastrtps_cpp/src/rmw_node.cpp

```diff
@@ -85,12 +85,34 @@ rmw_destroy_node(rmw_node_t * node)
     return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
 
   rmw_context_t * context = node->context;
-  rmw_ret_t ret = rmw_fastrtps_shared_cpp::__rmw_destroy_node(
+
+  rmw_ret_t ret = RMW_RET_OK;
+  rmw_error_state_t error_state;
+  rmw_ret_t inner_ret = rmw_fastrtps_shared_cpp::__rmw_destroy_node(
     eprosima_fastrtps_identifier, node);
   if (RMW_RET_OK != ret) {
-    return ret;
+    error_state = *rmw_get_error_state();
+    ret = inner_ret;
+    rmw_reset_error();
+  }
+
+  inner_ret = rmw_fastrtps_shared_cpp::decrement_context_impl_ref_count(context);
+  if (RMW_RET_OK != inner_ret) {
+    if (RMW_RET_OK != ret) {
+      RMW_SAFE_FWRITE_TO_STDERR(rmw_get_error_string().str);
+      RMW_SAFE_FWRITE_TO_STDERR(" during '" RCUTILS_STRINGIFY(__function__) "'\n");
+    } else {
+      error_state = *rmw_get_error_state();
```

Comment:
Is there an easy way to combine the two error states should there be multiple failures?

Reply:
As of today, there isn't. In general, we don't handle errors during error handling and/or finalization very well or at all. I opened an issue about this: https://github.com/ros2/rcutils/issues/269, to have an standard idiom for this.

For the moment, this makes sure the first error goes into the error state and any other errors that may follow do not get lost but do not obscure what is assumed to be the root cause of the error.